lib/ansible/modules/cloud/amazon/aws_s3_sync.py

```diff
@@ -0,0 +1,643 @@
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.aws.s3 import dict_to_s3_extra_args
+from ansible.module_utils.ec2 import AWSRetry
+from ansible.module_utils.common.dict_transformations import camel_dict_to_snake_dict
+from ansible.module_utils._text import to_bytes
+
+try:
+    from dateutil import tz
+    HAS_DATEUTIL = True
+except ImportError:
+    HAS_DATEUTIL = False
+
+try:
+    from hashlib import md5
+    HAS_MD5 = True
+except ImportError:
+    try:
+        from md5 import md5
+        HAS_MD5 = True
+    except ImportError:
+        HAS_MD5 = False
+
+try:
+    from botocore.exceptions import BotoCoreError, ClientError
+except ImportError:
+    pass    # Handled by AnsibleAWSModule
+
+
+class S3Syncer(object):
+    def __init__(self, module):
+        self.module = module
+        self.check_mode = module.check_mode
+        self.params = module.params
+        self.changed = False
+
+        params = self.params
+        self.file_args = module._module.load_file_common_arguments(params)
+        self.directory_args = copy(self.file_args)
+        self.directory_args['mode'] = params['directory_mode']
+
+        self.mode = params['direction']
+        self.bucket = params['bucket']
+        self.path = params['path']
+        self.prefix = params['prefix']
+        self.patterns = params['patterns']
+        self.excludes = params['excludes']
+        self.delete = params['delete']
+
+        self.mimetypes = mimetypes.MimeTypes()
+        if params['mime_override']:
+            self.mimetypes.encodings_map = params['mime_encodings_map']
+            self.mimetypes.types_map = ({}, params['mime_types_map'])
+        else:
+            self.mimetypes.encodings_map.update(params['mime_encodings_map'])
+            self.mimetypes.types_map[1].update(params['mime_types_map'])
+
+        try:
+            self.s3 = module.client('s3')
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg="Couldn't connect to S3")
+
+        self.objects = []
+
+    def _calculate_etag(self, entry):
+        if '-' in entry['s3']['e_tag']:
+            # Multi-part ETag; a hash of the hashes of each part.
+            parts = int(entry['s3']['e_tag'][1:-1].split('-')[1])
+            digests = []
+            with open(entry['path'], 'rb') as f:
+                for part_num in range(1, parts + 1):
+                    try:
+                        key_head = self.s3.head_object(Bucket=self.bucket, Key=entry['key'], PartNumber=part_num)
+                    except (BotoCoreError, ClientError) as e:
+                        self.module.fail_json_aws(e, msg="Couldn't get metadata for multi-part object %s" % entry['key'])
+                    digests.append(md5(f.read(int(key_head['ContentLength']))))
+            digest_squared = md5(b''.join(m.digest() for m in digests))
+            return '"{0}-{1}"'.format(digest_squared.hexdigest(), len(digests))
+
+        digest = md5()
```

Comment:
Any reason why `module.md5(entry['path'])` wasn't used here?

Reply:
This module uses AnsibleAWSModule, not AnsibleModule, so using AnsibleModule's md5() is a little ugly. We already needed to import and use md5 directly for the multipart digest, so calculating the single-part digest is a trivial amount of code.